railties/lib/rails/generators/rails/app/templates/Dockerfile.tt

```diff
@@ -9,6 +9,8 @@ WORKDIR /rails
 
 # Set production environment
 ENV RAILS_ENV="production" \
+    BUNDLE_DEPLOYMENT="1" \
```

Comment:
If I try this on an M1 mac, `docker buildx build . -t rails-demo` results in:

```
 > [build  5/10] RUN bundle _2.4.7_ install &&     bundle exec bootsnap precompile --gemfile:
#0 0.368 Your bundle only supports platforms ["arm64-darwin-22"] but your local platform
#0 0.368 is aarch64-linux. Add the current platform to the lockfile with
#0 0.368 `bundle lock --add-platform aarch64-linux` and try again.
```

Reply:
Yes, that is very much intended.

Reply:
Not the answer I was expecting to hear.  Just so we are clear, unless you are developing on a Linux development machine (while no long uncommon, certainly not the majority), `rails new` on 7.1 will generate a `Dockerfile` that can't be deployed **anywhere*** without developer intervention.

I personally think that this is a fine *option*, but I would argue against it as the default.

Note: the impact to me is minimal.  I can modify `fly launch` to detect this and add the necessary platform.

Reply:
Ok, so on modern-sh bundler (2.x mostly), you are supposed to list the platforms you intent to install on, via `bundle lock --add-platform`.

If you do your `rails new` from macOS, then yes, you'll have to run `bundle lock --add-platform aarch64-linux` to be able to build the docker image. And if you intend to deploy on x86 linx, you'll have  to run something like `bundle lock --add-platform x86-linux`.

That's how bundler works, and there's no way around it. If we don't set `BUNDLE_DEPLOYMENT`, bundler may do re-resolve the gems to find compatible version, which mean you can end up with totally different versions of your gems in the docker image, which is unacceptable.

So yeah that isn't ideal, but IMO the error message is quite clear and give you the command to run, so seem like the best we can do short from trying to assume which platform you gonna deploy to and to add the platfrom during `rails new`. But that seems very finicky to me.

Sorry, I'm quite used to this, I assumed you were too.

Reply:
Oh, and if this is added, dockerfile-rails will have a --no-lock flag.

Reply:
> dockerfile-rails will have a --no-lock flag.

You do you. But be ready for bug report of gem versions not matching the dev environment.

Reply:
What we can probably do, is to blindly add `x86-linux` and `aarch64-linux` on `rails new`, that should take care of the vast majority of use cases.

Reply:
To deploy on fly.io, we will need `x86_64-linux` listed as a platform.  And we've been providing Dockerfiles for quite some time, and no, we don't see bug reports of gem versions not matching the dev environment. 

And, in any case, if somebody explicitly adds a '--no-lock' flag and then sees a version mismatch, the recommendation will be something along the lines of "well, then, don't do that".  :-)

Reply:
> and no, we don't see bug reports of gem versions not matching the dev environment.

I believe you, but someone will get bit hard by this eventually. I also maintain a platform that build Docker images for thousands of apps, and I saw people bit by this.

It's well documented by bundler: https://bundler.io/guides/deploying.html, now it's your platform you do what you want.

Same with `yarn install` it's missing the `--frozen-lockfile` flag.

Reply:
> What we can probably do, is to blindly add x86-linux and aarch64-linux on rails new, that should take care of the vast majority of use cases.

I think we should do that, regardless of the Dockerfile. Most Rails apps are deployed to linux machines, no matter where the are developed. We already do that in our deployment platform.